src/app/digital/utils/ExpressionParser/GenerateInputTree.ts

```diff
@@ -76,27 +78,60 @@ const DefaultPrecedences: TokenType[] = ["|", "^", "&", "!", "("];
         rightRet = {index: index+1, tree: {kind: "leaf", ident: rightToken.name}};
     } else if (currentOp === "(") {
         if (index >= tokens.length)
-            throw new Error("Encountered Unmatched \"" + ops["("] + "\"");
+            throw new Error(`Encountered Unmatched "${ops["("]}"`);
         rightRet = generateInputTreeCore(tokens, ops, nextOpNum, index);
     } else {
         rightRet = generateInputTreeCore(tokens, ops, currentOpNum, index);
     }
     index = rightRet.index;
     if (currentOp === "(") {
         if (index >= tokens.length)
-            throw new Error("Encountered Unmatched \"" + ops["("] + "\"");
+            throw new Error(`Encountered Unmatched "${ops["("]}"`);
         if (tokens[index].type !== ")")
-            throw new Error("Encountered Unmatched \"" + ops["("] + "\"");
+            throw new Error(`Encountered Unmatched "${ops["("]}"`);
         rightRet.index += 1; // Incremented to skip the ")"
+        rightRet.final = true; // used to not combine gates in (a|b)|(c|d) for example
         return rightRet;
     }
 
     // The tree tree is created with the new node as the root and returned
     let tree: InputTree;
-    if (currentOp === "!")
-        tree = {kind: "unop", type: "!", child: rightRet.tree};
-    else if (currentOp === "|" || currentOp === "^" || currentOp === "&")
-        tree = {kind: "binop", type: currentOp, lChild: leftRet.tree, rChild: rightRet.tree};
+    if (currentOp === "!") {
+        const rTree = rightRet.tree;
+        if (rTree.kind === "binop" && !rTree.isNot) {
+            tree = rTree;
+            tree.isNot = true;
+        }
+        else {
+            tree = {kind: "unop", type: "!", child: rightRet.tree};
+        }
+    }
+    else if (currentOp === "|" || currentOp === "^" || currentOp === "&") {
+        const lTree = leftRet.tree;
+        const rTree = rightRet.tree;
+        // 7 + rTree = 8, that is why 7 is used
+        let childrenArray: InputTree[] = (lTree.kind === "binop"
+                             && lTree.type === currentOp
+                             && !lTree.isNot
+                             && !leftRet.final
+                             && lTree.children.length <= 7) ? lTree.children : [lTree];
```

Comment:
This is a lot of conditions, what exactly are you checking for? Might be better in a utility function like `isTreeBLAH()` especially since a lot of the conditions are repeated for `rTree` below